cpp/src/arrow/util/align_util.h

```diff
@@ -64,5 +66,23 @@ inline BitmapWordAlignParams BitmapWordAlign(const uint8_t* data, int64_t bit_of
   return p;
 }
 
+template <typename T>
+Status EnsureAlignment(std::shared_ptr<T> object, int64_t alignment,
```

Comment:
This method doesn't recurse into child arrays.  For example, a `StructArray` is going to have `object->children()` that will need to be visited as well.

Reply:
Now checking the children as well